### PR TITLE
basic dragDropFunctionality

### PR DESCRIPTION
**Pull request by Deepak123bharat**
In this PR I will add functionality to open project by drag and drop. I will also try to implemented functionality to add a new sample in an existing project by drag and drop.

### CODE REVIEW

**Comment by tomas-pluskal** [comment]
Hi @Deepak123bharat 
It's a good start, but I think we really need to add some logic for the file type recognition. Currently if I drop an mzML file to the app, it causes an error because it assumes it is an MZmine project file (=ZIP file).

One additional comment: on some platforms (like macOS) it is common to open files by dragging them to the application icon in the taskbar. To implement that, I think you can use this handler: https://docs.oracle.com/en/java/javase/13/docs/api/java.desktop/java/awt/Desktop.html#setOpenFileHandler(java.awt.desktop.OpenFilesHandler)

File: src/main/java/io/github/mzmine/gui/MZmineGUI.java [hunk 1]
```diff
@@ -111,6 +122,69 @@ public void start(Stage stage) {
...
+      @Override
+      public void handle(DragEvent event) {
+        Dragboard dragboard = event.getDragboard();
+        boolean success = false;
+        if (dragboard.hasFiles()) {
+          success = true;
+          for (File selectedFile:dragboard.getFiles()) {
+
+            final String extension = selectedFile.getName().substring(selectedFile.getName().lastIndexOf("."));
+            String[] RowDataFile = {".cdf",".nc",".mzData",".mzML",".mzXML"};
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/gui/MZmineGUI.java:148]: - change RowDataFile -> rawDataFile
- add ".raw" extension

File: src/main/java/io/github/mzmine/gui/MZmineGUI.java [hunk 2]
```diff
@@ -111,6 +122,69 @@ public void start(Stage stage) {
...
+    rootScene.setOnDragDropped(new EventHandler<DragEvent>() {
+      @Override
+      public void handle(DragEvent event) {
+        Dragboard dragboard = event.getDragboard();
+        boolean success = false;
+        if (dragboard.hasFiles()) {
+          success = true;
+          for (File selectedFile:dragboard.getFiles()) {
+
+            final String extension = selectedFile.getName().substring(selectedFile.getName().lastIndexOf("."));
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/gui/MZmineGUI.java:147]: What if the filename doesn't contain "."? 
Better to use existing utility method. There is FilenameUtils.getExtension() in Apache Commons IO, which is already included with MZmine.

File: src/main/java/io/github/mzmine/gui/MZmineGUI.java [hunk 3]
```diff
@@ -111,6 +122,69 @@ public void start(Stage stage) {
...
+            String[] RowDataFile = {".cdf",".nc",".mzData",".mzML",".mzXML"};
+            final Boolean isRawDataFile = Arrays.asList(RowDataFile).contains(extension);
+            final Boolean isMZmineProject = extension.equals(".mzmine");
+
+            String moduleClass = "";
+            if(isMZmineProject)
+            {
+              moduleClass = "io.github.mzmine.modules.io.projectload.ProjectLoadModule";
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/gui/MZmineGUI.java:155]: You don't need to use the class name as String and `Class.forName(..)`. The easiest way to refer a Java class is 
`Class<? extends MZmineRunnableModule> moduleJavaClass = ProjectLoadModule.class;`

File: src/main/java/io/github/mzmine/gui/MZmineGUI.java [hunk 4]
```diff
@@ -111,6 +122,69 @@ public void start(Stage stage) {
       e.consume();
     });
 
+    // Drag over surface
+    rootScene.setOnDragOver(new EventHandler<DragEvent>() {
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/gui/MZmineGUI.java:126]: Use the shorter lambda notation for event handlers, like 
```
rootScene.setOnDragOver(event -> {
   ... your code
});

```

File: src/main/java/io/github/mzmine/gui/MZmineGUI.java [hunk 5]
```diff
@@ -111,6 +122,69 @@ public void start(Stage stage) {
...
+      }
+    });
+    // Dropping over surface
+    rootScene.setOnDragDropped(new EventHandler<DragEvent>() {
+      @Override
+      public void handle(DragEvent event) {
+        Dragboard dragboard = event.getDragboard();
+        boolean success = false;
+        if (dragboard.hasFiles()) {
+          success = true;
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/gui/MZmineGUI.java:144]: "success" is a very generic name, try to name your variables in a way that when other people read your code they can immediately understand what is the purpose of the variable

**Comment by tomas-pluskal** [comment]
Thanks, I added a few comments. One issue is that your drag&drop code only works on the task controller TableView, but not on the Raw data files and Feature lists ListViews. That's because those controls have their own handlers for drag&drop for sorting the items. Can you try to solve this issue?

**Comment by Deepak123bharat** [comment]
I am trying to solve this.

File: src/main/java/io/github/mzmine/gui/MZmineGUI.java [hunk 6]
```diff
@@ -289,6 +314,52 @@ public static void showAboutWindow() {
...
+
+  public static void activateSetOnDragDropped(DragEvent event){
+    Dragboard dragboard = event.getDragboard();
+    boolean hasFileDropped = false;
+    if (dragboard.hasFiles()) {
+      hasFileDropped = true;
+      for (File selectedFile:dragboard.getFiles()) {
+
+        final String extension = FilenameUtils.getExtension(selectedFile.getName());
+        String[] rowDataFile = {"cdf","nc","mzData","mzML","mzXML","raw"};
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/gui/MZmineGUI.java:334]: rowDataFile -> rawDataFile

File: src/main/java/io/github/mzmine/gui/MZmineGUI.java [hunk 7]
```diff
@@ -289,6 +314,52 @@ public static void showAboutWindow() {
     });
   }
 
+  public static void activateSetOnDragOver(DragEvent event){
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/gui/MZmineGUI.java:317]: Please add JavaDoc comments to these new methods

File: src/main/java/io/github/mzmine/gui/MZmineGUI.java [hunk 8]
```diff
@@ -111,6 +127,15 @@ public void start(Stage stage) {
       e.consume();
     });
 
+    // Drag over surface
+    rootScene.setOnDragOver(event -> {
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/gui/MZmineGUI.java:131]: I think a better way to write this is 
`rootScene.setOnDragOver(MZmineGUI::activateSetOnDragOver);` 
(haven't tested it)

File: src/main/java/io/github/mzmine/gui/MZmineGUI.java [hunk 9]
```diff
@@ -19,13 +19,26 @@
...
+import java.io.FilenameFilter;
 import java.net.URL;
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.List;
 import java.util.Optional;
 import java.util.logging.Logger;
 import javax.annotation.Nonnull;
 
+import com.sun.istack.NotNull;
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/gui/MZmineGUI.java:32]: Not sure where this came from

**Comment by tomas-pluskal** [comment]
The code looks good now, but... it breaks the existing functionality. The user should be able to re-order the raw data files and feature lists by dragging them with a mouse. Please take a look at that issue.

**Comment by Deepak123bharat** [comment]
OK,  I will look into it.

File: src/main/java/io/github/mzmine/gui/MZmineGUI.java [hunk 10]
```diff
@@ -289,6 +305,63 @@ public static void showAboutWindow() {
...
+                MZmineCore.getConfiguration().getModuleParameters(moduleJavaClass);
+
+        if(isMZmineProject){
+          moduleParameters.getParameter(projectFile).setValue(selectedFile);
+        } else if (isRawDataFile){
+          File fileArray[] = { selectedFile };
+          moduleParameters.getParameter(fileNames).setValue(fileArray);
+        }
+
+        ParameterSet parametersCopy = moduleParameters.cloneParameterSet();
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/gui/MZmineGUI.java:357]: If you drop a file that doesn't have a recognized extension (e.g., .txt or .xls file), moduleParameters will be null here and cause a NullPointerException

**Comment by tomas-pluskal** [comment]
Looks great, thanks!

**Pull request merged**